### ISSUE TITLE
Testing firmware update via OTA

### ISSUE THREAD

**Issue by esc247** [roth-m/miioclient-mqtt#12]
Hi,

I have been trying to use the OTA firmware update method, as I burnt off the soldering point on the board. I cannot understand what to do to try this.

I installed both miio and python-miio, extracted the token from the device, but not sure what command to use.

Any help will be greatly appreciated.

**Comment by roth-m** [comment]
In the README there is the miio command I believe should be sent. But sorry, I never took time to test it.

**Comment by esc247** [comment]
I'm trying to test this, as I dont need the device (I got a CC2531). Happy to help test it if there was more info. I looked into the readme, but the command there doesnt work.

**Comment by roth-m** [comment]
Could you please try to issue a  miIO.ota_install command after the miIO.ota '{"mode":"normal", "install":"1", "app_url":"#url#", "file_md5":"#file_md5#"}' ?

**Comment by moodyblue** [comment]
Hi, I've bought a "Mi Smart Sensor Set" in a promo sometime ago. My plan was to use the sensors and trash the gateway, but then I saw this thread. So I've picked an old PC, installed Debian X86 on it, installed miiocli, and gave the following command:

```
hugo@debian:~$ miiocli device --ip 192.168.131.16 --token xxx info
Model: lumi.gateway.mieu01
Hardware version: Linux
Firmware version: 3.5.8_147
Network: {'localIp': '192.168.131.16', 'mask': '255.255.255.0', 'gw': '192.168.131.1'}
AP: {'ssid': 'Aveiro-Guest', 'bssid': '76:83:C2:9A:8B:04'}
```

This shows that miiocli is working, token is correct and my gateway is the mieu01 one. Then I gave the following command:

```
miiocli device --ip 192.168.131.16 --token xxx raw_command "miIO.ota \'{\"mode\":\"normal\", \"install\":\"1\", \"app_url\":\"https://github.com/roth-m/miioclient-mqtt/blob/master/firmware-mod/lummod_gw.bin\", \"file_md5\":\"525d3181fc3125eccb5e19540e8d771b\"}\'"
```

and waited some minutes. Then I've done a power recycle. Here's the situation now:

1) My android phone, using the "Mi Home" app, has access to the gateway
2) The gateway recognizes the sensors. I've tested one automation and it worked
3) telnet service is not enabled in the gateway
4) Command `miiocli device --ip 192.168.131.16 --token xxx info` no longer works (gives a timeout)

How to proceed from here ?

Edit: Looks like I've put a wrong md5 (got it from the description and not from the file itself), but apparently miiocli is no longer accessing the gateway (ping is ok)

**Comment by roth-m** [comment]
It looks like your gateway has changed its IP (dhcp) or token. Can you confirm by issuing a discover command?

**Comment by moodyblue** [comment]
In Mi Home app IP and token did not change. I have put the IP and token in OpenHAB (Xiaomi Mi IO binding) and it goes to online. In theory that binding is also capable of sending commands to the gateway, so it's an alternative to miiocli, but I don't know how to do it.

**Comment by roth-m** [comment]
The Xiaomi Mi IO binding will not work as the protocol used by this gateway (MIOT which is using MIIO commands) is not yet implemented (I think as of September 2020).
The firmware mod will shutdown the original miio-client after 10 minutes (from restart).
Can you ping the gateway or connect to it using ssh?

**Comment by moodyblue** [comment]
Don't know why, but `miiocli device --ip 192.168.131.16 --token xxx info` works again. So I've enabled a web server to put file `lummod_gw.bin` and gave the `miIO.ota` command. Waited 30mins and then power recycle the Mi gateway.

Now I can ping it, but I have no telnet or ssh access to it. Does your app open any ports so that I can inspect with nmap ?

**Comment by roth-m** [comment]
In the http server logs, do you see any request for the lummod_gw.bin file? In other words, does the gateway fetches the package?

Yes, it enables the ssh server (dropbear). You should be able to issue: ssh root@gateway_ip and login using the password "admin".

**Comment by moodyblue** [comment]
I need some time to investigate how to enable logging. I'm using a QNAP NAS with apache web server. apache log only contains boot messages. I'll try again in a few days.

**Comment by moodyblue** [comment]
I've enabled apache log. It shows download requests from my PC but not from the Mi Gateway. From my PC I can download using url `http://192.168.130.210:81/lummod_gw.bin`

I'm not a linux expert, maybe the `miIO.ota` syntax is not correct.

**Comment by roth-m** [comment]
Could you please try to issue the following command?

`miiocli device --ip 192.168.131.16 --token xxx raw_command "miIO.ota \'{\"mode\":\"normal\", \"install\":\"1\", \"app_url\":\"http://192.168.130.210:81/lummod_gw.bin\", \"file_md5\":\"525d3181fc3125eccb5e19540e8d771b\",\"proc\": \"dnld install\"}\'"`

**Comment by moodyblue** [comment]
Thanks but it did not work. One thing that I've noticed is that I can send several `info` commands to the gateway, but whenever I send a `raw_command miIO.ota ` command I need to power recycle the gateway before sending any more `miiocli` commands.

But even without a power recycle the gateway continues to be online in the Mi Home app.

**Comment by moodyblue** [comment]
I've tried, but the same symptom: I have no log that gateway has attempted to download the firmware, but in order to continue to be able to send `miiocli` commands I need to power recycle the gateway. But even without power recycle the gw appears online in the Mi Home app.

**Comment by roth-m** [comment]
So there is a need for a correct command to be sent for the firmware to be updated.....

**Comment by esc247** [comment]
Hi,

I have tried the above commands as well. Funnily, I can get the info of the device but after that when I try to update the device OTA with the command, I cannot seem to get the info of the device again unless I reboot the device.
________________________________

**Comment by roth-m** [comment]
Yeah I know. Sorry, I don't know any other way to update the firmware (trying to google around without succes).
Maybe try to issue a "miIO.get_ota_progress" command after the "miIO.ota" command ?!

**Comment by moodyblue** [comment]
I've tried several `miiocli` commands. info and some others are ok, but everything I try to run with raw_command produces a timeout and puts the gateway unavailable for a couple of minutes. Some even require a power recycle.
